src/astarte-client/types/event.ts

```diff
@@ -18,24 +18,190 @@
 */
 
 import type { AstarteDataType } from './dataType';
+import _ from 'lodash';
 
-interface AstarteDeviceConnectedEvent {
-  device_id: string;
-  timestamp: number;
-  event: {
-    type: 'device_connected';
-    device_ip_address: string;
-  };
+abstract class AstarteDeviceEvent {
+  private _deviceId: string;
+  private _timestamp: Date;
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  protected constructor(arg: any) {
+    if (!arg || !_.isPlainObject(arg)) {
+      throw Error('Invalid argument');
+    }
+
+    if (!arg.device_id || typeof arg.device_id !== 'string') {
+      throw Error('Invalid device id');
+    }
+
+    if (!arg.timestamp || typeof arg.timestamp !== 'number') {
+      throw Error('Invalid timestamp');
+    }
+
+    this._deviceId = arg.device_id;
+    this._timestamp = new Date(arg.timestamp);
+  }
+
+  get deviceId() : string {
+    return this._deviceId;
+  }
+
+  get timestamp() : Date {
+    return this._timestamp;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  static decode(arg: any): AstarteDeviceEvent | null {
+    return decodeAnyOf([
+      AstarteDeviceConnectedEvent.fromJSON,
+      AstarteDeviceDisconnectedEvent.fromJSON,
+      AstarteDeviceUnsetPropertyEvent.fromJSON,
+      AstarteDeviceIncomingDataEvent.fromJSON,
+    ], arg);
+  }
 }
 
-interface AstarteDeviceDisconnectedEvent {
-  device_id: string;
-  timestamp: number;
-  event: {
-    type: 'device_disconnected';
-  };
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type EventDecoder = (arg: any) => AstarteDeviceEvent;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function decodeAnyOf(decoders: EventDecoder[], value: any): AstarteDeviceEvent | null {
+  let decodedValue = null;
+
+  for (let i = 0; i < decoders.length && decodedValue === null; i += 1) {
+    try {
+      decodedValue = decoders[i](value);
+    } catch (err) {
+      // decoder not matching
+    }
+  }
+
+  return decodedValue;
```

Comment:
return inside of a try-catch block is considered a bad pattern, this is not the case, but when using `finally` previous return operations are ignored.

Reply:
iperative vs declarative here is just for early returns. We could make it declarative if the decoders would return `null` instead of throwing an exception

Reply:
I meant as a general note that you don't need to declare how to iterate an Array using indexes when you can choose a more straightforward `for of` loop